ui/src/app/views/workspace-share/component.spec.tsx

```diff
@@ -0,0 +1,140 @@
+    expect(wrapper.find(getSelectString(hermioneRole)).text()).toEqual(fp.capitalize(WorkspaceAccessLevel[hermioneRole.role]));
+    expect(wrapper.find(getSelectString(ronRole)).text()).toEqual(fp.capitalize(WorkspaceAccessLevel[ronRole.role]));
+    expect(wrapper.find(getSelectString(lunaRole)).length).toBe(0);
+  });
+
+  it('removes user correctly', () => {
+    const wrapper = component();
+    wrapper.find('[data-test-id="remove-collab-ron.weasley@hogwarts.edu"]').first().simulate('click');
+    const expectedNames = [harry, hermione].map(u => u.givenName + ' ' + u.familyName);
+    expect(wrapper.find('[data-test-id="collab-user-name"]').map(el => el.text())).toEqual(expectedNames);
+  });
+
+  it('adds user correctly', async() => {
+    let clock = Lolex.install({shouldAdvanceTime: true});
+    const wrapper = component();
+    wrapper.find('[data-test-id="search"]').simulate('change', {target: {value: 'luna'}});
+    clock.tick(401);
+    await waitOneTickAndUpdate(wrapper);
+    wrapper.update();
+    clock.uninstall();
+    wrapper.find('[data-test-id="add-collab-luna.lovegood@hogwarts.edu"]').first().simulate('click');
+    const expectedNames = [harry, hermione, ron, luna].map(u => u.givenName + ' ' + u.familyName);
+    expect(wrapper.find('[data-test-id="collab-user-name"]').map(el => el.text())).toEqual(expectedNames);
+  });
+
+  it('does not allow self-removal', () => {
+    const wrapper = component();
+    const dataString = '[data-test-id="remove-collab-harry.potter@hogwarts.edu"]';
+    expect(wrapper.find(dataString).length).toBe(0);
+  });
+
+  it('does not allow self role change', () => {
+    const wrapper = component();
+    expect(wrapper.find('[data-test-id="harry.potter@hogwarts.edu-user-role"]').first()
+      .props()['isDisabled']).toBe(true);
+  });
+
+  it('saves acl correctly after changes made', async() => {
+    let clock = Lolex.install({shouldAdvanceTime: true});
+    const wrapper = component();
+    wrapper.find('[data-test-id="search"]').simulate('change', {target: {value: 'luna'}});
+    clock.tick(401);
+    await waitOneTickAndUpdate(wrapper);
+    wrapper.update();
+    clock.uninstall();
+    // add luna to acl
+    wrapper.find('[data-test-id="add-collab-luna.lovegood@hogwarts.edu"]').first()
+      .simulate('click');
+    // remove ron from acl
+    wrapper.find('[data-test-id="remove-collab-ron.weasley@hogwarts.edu"]').first()
+      .simulate('click');
+    // change hermione's access to owner
+    let selectComponent = wrapper.find('[data-test-id="hermione.granger@hogwarts.edu-user-role"]')
+      .find(Select);
+    selectComponent.instance().setState({menuIsOpen: true});
+    wrapper.update();
+    wrapper.find('.hermionegrangerhogwartsedu-user-role__option')
+      .findWhere(n => n.text() === fp.capitalize(WorkspaceAccessLevel[WorkspaceAccessLevel.OWNER]))
+      .simulate('click');
+    wrapper.update();
+    const spy = jest.spyOn(workspacesApi(), 'shareWorkspace');
```

Comment:
nit: it would be better to just check the UI after, rather than spying on this method, since you went through the effort of implementing the stub as well. Or does that part not work (this is the reload thing?)

Reply:
The reload thing definitely an issue on the workspace page. And on the workspace list page there would be nothing to check on the UI (since the page doesn't display workspaces ACLs). So I thought best way to test was to make sure this method got called with the correct request.